provisioning/device/samples/solutions/BestPracticeSampleX509/BestPracticeSampleX509.cs

```diff
@@ -0,0 +1,533 @@
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Azure.Devices.Provisioning.Client.Transport;
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Client.Samples;
+using Microsoft.Azure.Devices.Client.Exceptions;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Samples
+{
+    /// <summary>
+    /// Demonstrates how to register a device with the device provisioning service using a X.509 leaf certificate, and then
+    /// use the registration information to authenticate to IoT Hub.
+    /// </summary>
+    internal class BestPracticeSampleX509
+    {
+        private static readonly Random s_randomGenerator = new();
+        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(15);
+
+        private static readonly SemaphoreSlim s_initSemaphore = new(1, 1);
+        private readonly TransportType _transportType;
+        private static readonly ClientOptions s_clientOptions = new() { SdkAssignsMessageId = SdkAssignsMessageId.WhenUnset };
+
+        private readonly ILogger _logger;
+
+        // An UnauthorizedException is handled in the connection status change handler through its corresponding status change event.
+        // We will ignore this exception when thrown by client API operations.
+        private static readonly HashSet<Type> s_exceptionsToBeRetried = new()
+        {
+            // Unauthorized exception conditions are handled by the ConnectionStatusChangeHandler in this sample and the sample will try
+            // to reconnect indefinitely, so don't give up on an operation that sees this exception.
+            typeof(UnauthorizedException),
+        };
+
+        private readonly IRetryPolicy _customRetryPolicy;
+
+        // Mark these fields as volatile so that their latest values are referenced.
+        private static volatile DeviceClient s_deviceClient;
+
+        private static volatile ConnectionStatus s_connectionStatus = ConnectionStatus.Disconnected;
+
+        private static CancellationTokenSource s_appCancellation;
+
+        // A safe initial value for caching the twin desired properties version is 1, so the client
+        // will process all previous property change requests and initialize the device application
+        // after which this version will be updated to that, so we have a high water mark of which version number
+        // has been processed.
+        private static long s_localDesiredPropertyVersion = 1;
+
+        private readonly Parameters _parameters;
+
+        private readonly string _globalEndpoint;
+        private readonly string _idScope;
+        private readonly string _certificateName;
+        private string _certificatePassword;
+        private DeviceAuthenticationWithX509Certificate _auth;
+        private string _assignedIoTHub;
+
+        public BestPracticeSampleX509(string globalEndpoint, string idScope, string certificateName, string certificatePassword, TransportType transportType, ILogger logger, Parameters parameters)
```

Comment:
maybe debatable, if "parameters" are send, maybe there is no need to send the individual parameter (globalEndpoint, idScope, certificateName ....)
```suggestion
        public BestPracticeSampleX509(ILogger logger, Parameters parameters)
```